tfhe/src/high_level_api/integers/unsigned/inner.rs

```diff
@@ -171,20 +195,26 @@ impl RadixCiphertext {
             Self::Cpu(cpu_ct) => cpu_ct,
             #[cfg(feature = "gpu")]
             Self::Cuda(ct) => {
-                with_thread_local_cuda_streams(|streams| ct.to_radix_ciphertext(streams))
+                with_thread_local_cuda_streams_for_gpu_indexes(ct.gpu_indexes(), |streams| {
+                    ct.to_radix_ciphertext(streams)
+                })
             }
         }
     }
 
     #[cfg(feature = "gpu")]
-    pub(crate) fn into_gpu(self) -> crate::integer::gpu::ciphertext::CudaUnsignedRadixCiphertext {
+    pub(crate) fn into_gpu(self, streams: &CudaStreams) -> CudaUnsignedRadixCiphertext {
         match self {
-            Self::Cpu(cpu_ct) => with_thread_local_cuda_streams(|streams| {
-                crate::integer::gpu::ciphertext::CudaUnsignedRadixCiphertext::from_radix_ciphertext(
-                    &cpu_ct, streams,
-                )
-            }),
-            Self::Cuda(ct) => ct,
+            Self::Cpu(cpu_ct) => {
+                CudaUnsignedRadixCiphertext::from_radix_ciphertext(&cpu_ct, streams)
+            }
+            Self::Cuda(ct) => {
+                if ct.gpu_indexes() == streams.gpu_indexes() {
+                    ct
+                } else {
+                    ct.duplicate(streams)
+                }
```

Comment:
I've seen this small pattern in a few places, could it be a ct.duplicate_if_needed(streams) ?

Reply:
That could be added yes